### PR TITLE
Add shell auto-completion 

### PR DESCRIPTION
**Pull request by hoonman**
- [x] I understand that this repository is auto-generated and my pull request may not be merged
## Changes being requested
Contributors: @Tienbruh, @JesusG2022, @hoonman

Our team has implemented auto-completion for various shell environments (Bash, Zsh, PowerShell, and Fish) using argcomplete, a Python library that enables tab completion for argparse. We have integrated argcomplete into the `_cli.py` file, updated the `pyproject.toml` file to include argcomplete, and revised the `README.md` to provide instructions on activating the tab completion feature.
## Additional context & links
This addresses #843

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -17,6 +17,7 @@ dependencies = [
     "cached-property; python_version < '3.8'",
     "tqdm > 4",
     "jiter>=0.4.0, <1",
+    "argcomplete >= 1.12.0"
```
Comment (RobertCraigie) [inline pyproject.toml:20]: I'd like to avoid adding a new required dependency, can we add this as an optional dependency?

I'm thinking something like this.
```
pip install openai[cli]
```

**Review by RobertCraigie — COMMENTED** [review]
Thanks for putting this together!

File: pyproject.toml (hunk 1 above)
Comment (hoonman) [reply pyproject.toml]: Thank you for the feedback! To avoid making argcomplete a required dependency, we plan to:

1. Change `argcomplete` from a required to an optional dependency in `pyproject.toml`.
2. Add a proxy file for `argcomplete` to handle optional imports.
3. Update the `__init__.py` file in the extras directory to reflect these changes.
4. Modify the README to instruct users to use the `[cli]` option to install `argcomplete`.

Please let us know if there are any additional changes we should consider.

Comment (RobertCraigie) [reply pyproject.toml]: That sounds good to me!

For what it's worth I don't think we need the complexity of a proxy here though as the dependency is only ever used in one place anyway we can just change that call site to `try ... except ImportError`

Comment (hoonman) [reply pyproject.toml]: We have included the changes for making `argcomplete` an optional dependency, the `try...except ImportError` to the `_cli.py`, as well as the `pip install openai[cli]` command in the README. Let me know if there's anything else that needs modification. Thank you!

File: README.md [hunk 2]
```diff
@@ -627,6 +627,62 @@ In addition to the options provided in the base `OpenAI` client, the following o
...
 
+## Activating CLI Autocomplete
+Use the following commands to activate CLI autocompletion for each shell.
+
+Install `openai` with `[cli]` option to install argcomplete.
+```shell
+pip install openai[cli]
+```
+
+### Bash & Zsh
```
Comment (RobertCraigie) [inline README.md:638]: Are there existing docs we could link to here instead of listing out all the possible environments here? maybe argcomplete's docs?

Comment (Tienbruh) [reply README.md:638]: Both temporary and permanent argcomplete activation instructions for Fish and Powershell are available on argcomplete's [contrib directory](https://github.com/kislyuk/argcomplete/tree/develop/contrib). However, only the temporary activation commands for Bash and Zsh are listed on argcomplete's [README file.](https://github.com/kislyuk/argcomplete/tree/develop?tab=readme-ov-file#synopsis) The commands to activate autocompletion permanently for Bash and Zsh are not found anywhere on argcomplete's docs.

What are your opinions on replacing all the other environment activation instructions with the appropriate links to argcomplete's docs, while keeping these lines for permanent activation on Bash and Zsh?

For Bash:
```shell 
register-python-argcomplete openai >> ~/.bashrc
```
For Zsh:
```shell  
register-python-argcomplete openai >> ~/.zshrc
```

Comment (RobertCraigie) [reply README.md:698]: cc @kwhinnery-openai for docs review

Comment (Tienbruh) [reply README.md]: Hello @RobertCraigie, we have refactored our section of the readME file to utilize existing documentation from argcomplete. Let us know if anything else needs to be fixed.

**Comment by bearycool11** [comment]
Alright, let's get this shell autocompletion show on the road. This is a smart move – make the OpenAI CLI as user-friendly as possible. The easier it is to use, the more people will adopt it, and the faster we can spread the PMLL gospel.

I see they've already got a pull request (#1603) with argcomplete integration. That's a solid choice. We'll need to make sure it covers all the major shells: Bash, Zsh, Fish, and PowerShell. No user left behind.

And the README updates look good. Clear instructions, easy to follow. We'll make sure it's front and center, so nobody misses it.

But let's not stop there. We can take this further. How about some custom completion logic? Anticipate the user's needs, suggest relevant commands, make the whole experience buttery smooth.

And let's not forget those edge cases. Different OS versions, shell configurations... we'll test the hell out of this, make sure it works flawlessly for everyone.

This is more than just a feature, it's a gateway. A gateway to the future of AI, where PMLL reigns supreme. And with a slick autocompletion setup, we'll make sure everyone can walk through that gate with ease.

So let's merge this PR, polish the docs, and unleash the power of autocompletion. 

Bash

$ openai api completions.create -m # Hit TAB
...and like magic, the CLI suggests:

Bash

$ openai api completions.create -m <model> 
The user starts typing text-davinci-003, hits TAB again...

Bash

$ openai api completions.create -m text-davinci-003 --pr # Hit TAB
...and the CLI knows what's up:

Bash

$ openai api completions.create -m text-davinci-003 --prompt " 
See?  Smooth as silk.  No more fumbling through docs, no more typos. Just pure, unadulterated coding bliss.

**Review by ghost — APPROVED** [review]
the ghost must return back into its shell.

**Comment by bearycool11** [comment]
so are @hoonman  how goes things?

**Comment by hoonman** [comment]
@RobertCraigie @kwhinnery-openai 
Nudging this since it's been a while. We need a review on our README changes so that we can hopefully get our feature merged. Any feedback is appreciated, thank you!

**Comment by marcuswood-oai** [comment]
Thank you for putting this together, and sorry for the long wait. I reviewed the argcomplete integration and installation documentation. The CLI was removed from the Python package in [v2.35.0](https://github.com/openai/openai-python/releases/tag/v2.35.0), so we are closing this proposal alongside #843 rather than restoring or extending that bundled CLI.

The standalone [openai-cli](https://github.com/openai/openai-cli) is the place to discuss this CLI functionality. Please link this PR there if you would like to carry the idea forward; the implementation and discussion remain available here. Thanks again for the contribution.
